### ISSUE TITLE
Current docker image doesn't run on ARM64

### ISSUE THREAD

**Issue by ThaMattie** [PlexRipper/PlexRipper#137]
It would be nice if the image on dockerhub would also run on ARM64 architecture.
Currently the image on dockerhub is x86-64 only.

**Comment by jinglebinkie** [comment]
This would be cool!  I'm trying to run this on Kubernetes on Raspberry, but figured out just too late the docker images is not ready for arm ... My code is ready to deploy ;-) .. 
Keep us posted please @JasonLandbridge !

**Comment by JasonLandbridge** [comment]
Hi there, sure I will be looking into this! From the quick research I did it shouldn't be too difficult. To help out with the process, can you guys list your hardware components (CPU, RAM, GPU, etc) and what OS you are running on it. 

Do you guys happen to have experience with creating docker containers for ARM64 architecture?

**Comment by ThaMattie** [comment]
Hi I'm running Armbian on a Pine64+ with 2gb memory:

The original Pine A64 boards released in 2016 are powered by the Allwinner A64 system-on-chip. It features a 1.2 GHz Quad-Core ARM Cortex-A53 64-Bit Processor, an ARM Mali 400 MP2 graphics processor unit, one HDMI 1.4a port, one MicroSD slot, two USB 2.0 ports and a 100 Megabit Ethernet port. The A64 board has only 512 megabytes of RAM, the 1 GB and 2 GB versions are labeled "Pine A64+".[10] While the 512 MB model only works with Arch Linux and Debian GNU/Linux distributions such as Armbian or DietPi,[11] the A64+ with more memory can also run other operating systems including Android, Remix OS, Windows 10,[12] FreeBSD,[13] and Ubuntu.[14] Optional eMMC storage modules can be plugged into special headers on the board.


Unfortunately I have no experience in building any kind of docker image yet.

**Comment by CodeSpent** [comment]
I've validated & integrated `arm64` support & images in a fork. So far I haven't seen any issues, so pretty confident in this. Will be working with maintainer to get this merged in a cohesive way.

Currently building the image natively from `arm64` system, however, will need an actual build system for CI/CD for maintaining across releases. I'm looking at a `buildx` implementation.

In the meantime, if you have access to an `arm64` system then building this image isn't too difficult;

1. Fork & Clone the repository.
2. Navigate to the project directory in your terminal.
3. Ensure Docker is installed & Docker daemon is running.
4. In the project root, build an image for arm64 using `docker build . --platform=linux/arm64 --no-cache --tag plexripper/plexripper:arm`
5a. Once the build is complete, run this newly tagged image `docker run -d plexripper/plexripper:arm`
5b. You can also use Docker Desktop to start this image as a new container.

Bare in mind with this workaround:
- You will need to manually pull the repository changes between versions & rebuild until we work out a CI/CD solution for multi-arch support.
- You may need to specify `--platform=linux/arm64` if you've previously pulled the official `plexripper/plexripper` image due to namespace clashing. This is why I tagged the image with `:arm` to avoid my own confusion as well.


I will be working with the maintainer to get this build process into their CI/CD pipeline so we can enjoy arm64 support in future releases. :)

**Comment by ThaMattie** [comment]
I tried building it from Portainer real quick, but it errors out:
![image](https://github.com/PlexRipper/PlexRipper/assets/12159972/101c2470-fdfe-47af-a94d-9634d2b7cad6)

It did create some images, but no clue how to go from there
![image](https://github.com/PlexRipper/PlexRipper/assets/12159972/8b5a5416-cef7-44ab-99a2-36fe24d9a1df)

**Comment by CodeSpent** [comment]
> I tried building it from Portainer real quick, but it errors out: ![image](https://user-images.githubusercontent.com/12159972/264163770-101c2470-fdfe-47af-a94d-9634d2b7cad6.png)
> 
> It did create some images, but no clue how to go from there ![image](https://user-images.githubusercontent.com/12159972/264164458-8b5a5416-cef7-44ab-99a2-36fe24d9a1df.png)

I've not used Portainer before so I can't attest to how they're handling here, but that error is exactly what you'd get if you ran the entrypoint script in the wrong architecture.

Double check the architecture of the images being created.

Its also possible that Portainer is using their own custom images for testing to report to the UI here, and if that's the case they may be using the wrong arch in images either by oversight or a lack of manifest in this repo (currently working on).

Is Portainer hosted on your arm64 system? Could you give me some more details on the implementation & system details so I can emulate your environment?

**Comment by ThaMattie** [comment]
> Double check the architecture of the images being created.
I wouldn't know how...

> Is Portainer hosted on your arm64 system? Could you give me some more details on the implementation & system details so I can emulate your environment?

I'm running this on a RPI 4, 8GB which is ARM64 indeed. I've installed:

- the official Raspbian 64bit lite (no desktop) image
- latest OpenMediaVault
- OpenMediaVault compose plugin (it includes Docker)
- Portainer via compose file
- Git

Then in Portainer I created a local environment and tried the Build Image from URL (as per Portainer documentation) using both the official PlexRipper and your fork as source.

I assumed it used the system architecture for building, but I don't know how the process works. Documentation says it clones the repo and then executes the dockerfile

**Comment by CodeSpent** [comment]
I don't use Portainer, but I'm unable to reproduce this. That does seem like it isn't being built for the proper platform, though, as that's the same error I get trying to run the Plexripper app with Rosetta as an arm64 build.

I'd try and build this outside of Portainer using the commandline from your arm64 system & see if you get the same issue there.

**Comment by ThaMattie** [comment]
I did the build outside of portainer, and then I'm able to run it!
Hope it gets added to the build pipeline soon, I try to have everything automated as much as possible

**Comment by frosit** [comment]
Any assistance required with this issue? (I also read #254) 
Just spoke to @JasonLandbridge about trying to share some skills on each others' projects and I have experience with the topics/technologies discussed in this issue so, happy to help.

I also have quite the variety of hardware available here, a lot of boards and a few other ARM based devices. My Pi's 3&4 for example aren't currently used so together with my server they could easily be setup to test/debug multi-arch builds. Which I might do anyway for other contributions to this project when I become more familiar with it.

**Comment by CodeSpent** [comment]
> Any assistance required with this issue? (I also read #254) Just spoke to @JasonLandbridge about trying to share some skills on each others' projects and I have experience with the topics/technologies discussed in this issue so, happy to help.
> 
> I also have quite the variety of hardware available here, a lot of boards and a few other ARM based devices. My Pi's 3&4 for example aren't currently used so together with my server they could easily be setup to test/debug multi-arch builds. Which I might do anyway for other contributions to this project when I become more familiar with it.

Absolutely. My bandwidth just isn't there to emulate the CI for this project in order to really refine it. Essentially all that is needed is a task in the pipeline that builds for arm64 using buildx or something similar.

I can absolutely provide direction, but I've been swamped for time to extend the necessary effort.

**Comment by ThaMattie** [comment]
> > Double check the architecture of the images being created.
> 
> I wouldn't know how...
> 
> > Is Portainer hosted on your arm64 system? Could you give me some more details on the implementation & system details so I can emulate your environment?
> 
> I'm running this on a RPI 4, 8GB which is ARM64 indeed. I've installed:
> 
>     * the official Raspbian 64bit lite (no desktop) image
> 
>     * latest OpenMediaVault
> 
>     * OpenMediaVault compose plugin (it includes Docker)
> 
>     * Portainer Business Edition via compose file
> 
>     * Git
> 
> 
> Then in Portainer I created a local environment and tried the Build Image from URL (as per Portainer documentation) using both the official PlexRipper and your fork as source.
> 
> I assumed it used the system architecture for building, but I don't know how the process works. Documentation says it clones the repo and then executes the dockerfile

Just to inform: I've had issues with other repo's not being build correctly through Portainer in this setup, and it is pobably something OpenMediaVault does with the Docker or Portainer install. When I test with Armbian (on pine64) or Raspbian (on rpi4) and installing Docker and Portainer without first installing OpenMediaVault, I can actually build PlexRipper straigth from this repo in Portainer

**Comment by frosit** [comment]
@ThaMattie that's the golden comment, thank you sir!

I'm very new to this project and this comment really helped me connect the dots in all the issues and current project status, so if I'm correct, teason why our image doesn't run on ARM is because we don't build ARM builds in our pipeline, we have to extend it and build using buildx and as stated in #254 @CodeSpent already has work in progress on this topic.

It makes sense that pulling the image from the registry using for example portainer wouldn't work on ARM platforms as ARM images are not available in the registry. Our pipeline simply does not build for ARM (yet). Building it yourself from sources works and I can confirm that it does (thanks all!). 

I just build the dev branch on my Pi4 using buildx, pushed the image to docker hub and pulled/deployed it as container using portainer. [FrosIT/PlexRipper-ARM](https://hub.docker.com/repository/docker/frosit/plexripper-arm). If you want to test it, make sure to setup the mounts properly as there's no... app template? (maybe a nice addition)

I'll see if I can combine the build stages into images of both flavours and push them as the one, the image above is arm only but I think this won't take long. I'll talk to @JasonLandbridge if we can push images manually for now so we don't have to wait for the update of our pipeline.

__About OpenMediaVault issues__
I've been running Openmediavault for a few years myself and @ThaMattie may have a point there. Unfortunately my server has a dead power supply so I can't currently test but maybe it's best to reopen this in a separate issue. They don't seem related.

**Comment by frosit** [comment]
We added multi-arch images to Docker hub under the __test__ tag for now. It would be greatly appreciated if people could report back feedback on using these.

To use these, use: `plexripper/plexripper:test` for image. 2 flavours of ARM we're added which would be sufficient in most cases. 

[PlexRipper/PlexRipper:test](https://hub.docker.com/layers/plexripper/plexripper/test/images/sha256-2af0ad85d1ee77632a1595c6b71f038d46a283e49f18da655f56f16a519a7e9e?context=explore)

@CodeSpent I'm currently not planning to update our workflow for buildx, these we're manually pushed. In case you find the time to get to a PR that would be great. I will direct my focus to other things for now.

**Comment by ThaMattie** [comment]
> It would be greatly appreciated if people could report back feedback on using these.

I just tested it on rpi4 and can confirm it works.
I copied the db from my own (older) build, and it migrated successfully. Performed a test rip as well and everything was as expected.

**Comment by jinglebinkie** [comment]
I tested today, created a small k8s deployment for microk8s on my RPI 4, with NAS attached volumes. Also I can confirm it works. 
Maybe 1 suggestion, after I've setup everything, (following the config/setup wizard), after re-opeing the page, it still asks if i want to set it up again of skip setup. Maybe add a detection if config volume is available and populated , then skip wizard page. Thanks for this !

**Comment by frosit** [comment]
Good suggestion, that's one for @JasonLandbridge to consider. Thanks for testing all! Will push the images to dev and latest tomorrow!

**Comment by JasonLandbridge** [comment]
> Maybe 1 suggestion, after I've setup everything, (following the config/setup wizard), after re-opeing the page, it still asks if i want to set it up again of skip setup. Maybe add a detection if config volume is available and populated , then skip wizard page. Thanks for this !

@jinglebinkie Are you talking about this? 

![image](https://github.com/PlexRipper/PlexRipper/assets/15127381/ca0fa0f5-b703-4456-ac1b-dd6a47ed8b5b)

Clicking on the skip setup button will hide it

**Comment by roelo** [comment]
Indeed, exactly this , but it was setup, hence the notification (and text in it) are not needed. Skipping indeed let me continue 👍

**Comment by ThaMattie** [comment]
Ah yes, I noticed that as well, when I used the new test image but already had data from a previous image

**Comment by JasonLandbridge** [comment]
This should now work with the latest dev build. Please see: https://github.com/PlexRipper/PlexRipper/issues/363
